depend: limit DLL parent path preservation to site-packages

rokm opened the pull request:
Limit the DLL parent path preservation only to files that are collected from site-packages directories, as returned by `site` `getsitepackages()`  and `getusersitepackages()`.

The initial implementation using all paths from `sys.path` has proven to be problematic due to our lack of control over `sys.path`. The latter may end up containing a drive root, causing all DLLs collected from that drive that are not explicitly rooted in some other directory also in `sys.path` to be collected with its whole directory structure preserved.

The DLL parent path preservation is primarily intended for PyPI wheels that ship and install their own copies of DLLs along with the package. As both are typically installed into site-packages directory, limiting the DLL parent path preservation to only site-packages directories should be both sufficient for our use case and safe, as it has no effect on the outside paths.

Fixes #7155.

rokm committed:
depend: limit DLL parent path preservation to site-packages

Limit the DLL parent path preservation only to files that are
collected from site-packages directories, as returned by `site`
`getsitepackages()`  and `getusersitepackages()`.

The initial implementation using all paths from `sys.path` has
proven to be problematic due to our lack of control over `sys.path`.
The latter may end up containing a drive root, causing all DLLs
collected from that drive that are not explicitly rooted in some
other directory also in `sys.path` to be collected with its whole
directory structure preserved.

The DLL parent path preservation is primarily intended for PyPI
wheels that ship and install their own copies of DLLs along with
the package. As both are typically installed into site-packages
directory, limiting the DLL parent path preservation to only
site-packages directories should be both sufficient for our use
case and safe, as it has no effect on the outside paths.

rokm commented:
The problem in #7155 is that the drive root ended up in `sys.path`, and that our initial implementation of the DLL path preservation used `sys.path` but explicitly excluded `sys.prefix` and `sys.base_prefix`. The latter was done in order to avoid explicitly dealing with conda's DLL directories. A straight-forward fix for this particular scenario would be to leave `sys.prefix` / `sys.base_prefix` among the considered paths, and explicitly adding the conda's DLL directory as well.

However, if the drive root had been `c:\`, the same issue would arise for any DLL collected from `c:\Windows` (we sometimes pick VC runtime DLLs from there) or from MSVC or Windows SDK installations in `c:\Program Files\...` (we pick `ucrtbase.dll` and `ms-api-*` DLLs from there, if the installation is in `PATH`). So after some thought, I think the sanest way is to limit the DLL path preservation only to site-packages directories, because it is trying to accommodate the DLLs collected from PyPI wheels, whose directory layout needs to be preserved for one reason or another.

bwoodsend commented:
Hmm, I'm not particularly fond of having different behaviours for a package installed with `pip install .` vs `pip install -e .`. This feels to me like another case for using `[i.files for i in importlib.metadata.distributions()]` as the determinant for whether a file originated from a pip package or otherwise and in this case, it would give us the destination path for free. We shied away from that before though because we expected that it would take a while on full blown Conda. I'm rather leaning towards doing it anyway though. What do you think?

rokm commented:
> Hmm, I'm not particularly fond of having different behaviours for a package installed with `pip install .` vs `pip install -e .`. This feels to me like another case for using `[i.files for i in importlib.metadata.distributions()]` as the determinant for whether a file originated from a pip package or otherwise and in this case, it would give us the destination path for free. We shied away from that before though because we expected that it would take a while on full blown Conda. I'm rather leaning towards doing it anyway though. What do you think?

Hmm, that might indeed be a better approach for this problem. I take it the hesitation to use it was due to the fact that it is a brute-force scan through all available distributions, as opposed to the look-up of the path when it is needed?

bwoodsend commented:
Yeah, pretty much. Iterating through `importlib.metadata.distributions()` used to take a while but it might have sped up since. I remember it taking about 10 seconds last time but now it's almost instant. I don't think that my environment has shrunk significantly since then.

rokm committed:
depend: DLL parent path preservation: work around for pywin32

The `win32uiole` extension is linked agains `win32ui` extension,
but due to the `sys.path` manipulation, the ` win32uiole` is a top
level module, and is collected to top-level directory. Therefore,
the same treatment needs to be applied to the `win32ui` - we should
not preserve its parent directory structure. Apply the same exception
to the whole `site-packages/pythonwin` directory...

rokm commented:
Hmm, I wonder if that will be a problem for our CI... but building the mapping of all files to their parent distributions could be useful elsewhere, too: for example in license collector; or if we added an option to automatically collect metadata for all dists from which we collect something; etc.

---

But at any rate, looks like this particular area (DLL parent path preservation) won't be possible without some additional hard-coded work-arounds. In this case, for `pywin32`, where `win32uiole.pyd` is linked against `win32ui.pyd`. But because `pywin32` adds whole bunch of its directories to `sys.path`, those are actually top-level modules and we collect them as such. So we actually must not preserve the DLL parent path for `win32ui.pyd` when we pick it up via dependency analysis. (And the current design has no problems with this because it is based on `sys.path`, so that extension is located directly under one of DLL preservation directories).
